go/service/chat_local.go

```diff
@@ -89,58 +175,133 @@ func (h *chatLocalHandler) CompleteAndCanonicalizeTlfName(ctx context.Context, t
+		resp, err := h.boxer.tlf.CryptKeys(ctx, criteria.TlfName)
 		if err != nil {
-			return ids, err
+			return nil, err
 		}
-		for _, conv := range iview.Conversations {
-			if conv.Metadata.IdTriple.Tlfid.Eq(tlfID) {
-				// TODO: check topic name and topic ID here when we support multiple
-				// topics per TLF
-				ids = append(ids, conv.Metadata.ConversationID)
+		tlfIDb, err := hex.DecodeString(string(resp.TlfID))
+		if err != nil {
+			return nil, err
+		}
+		conversationsRemote, err := h.remoteClient().GetInboxByTLFIDRemote(ctx, tlfIDb)
+		if err != nil {
+			return nil, err
+		}
+		for _, cr := range conversationsRemote {
+			info, _, err := h.getConversationInfo(ctx, cr.Metadata.ConversationID)
+			if err != nil {
+				return nil, err
+			}
+			if info.TlfName != criteria.TlfName {
+				// check again using signed information to make sure it's the correct
+				// conversation
+				return nil, fmt.Errorf("Unexpected data is returned from server. We asked for %v, but got conversation for %v. TODO: handle tlfName changes properly for SBS case", criteria.TlfName, info.TlfName)
 			}
+			appendMaybe(info)
 		}
+	} else { // slow case: tlfName not specified; need to search through inbox
+		ipagination := &chat1.Pagination{Num: 20}
+	getinbox:
+		for i := 0; i < 10000; /* in case we have a server bug */ i++ {
+			iview, err := h.GetInboxLocal(ctx, ipagination)
+			if err != nil {
+				return conversations, err
+			}
+			for _, conv := range iview.Conversations {
+				info, _, err := h.getConversationInfo(ctx, conv.Metadata.ConversationID)
+				if err != nil {
+					return conversations, err
+				}
+				appendMaybe(info)
+			}
 
-		if iview.Pagination == nil || iview.Pagination.Last {
-			break getinbox
-		} else {
-			ipagination = iview.Pagination
+			if iview.Pagination == nil || iview.Pagination.Last {
+				break getinbox
+			} else {
+				ipagination = iview.Pagination
+			}
 		}
 	}
 
-	id, err := h.NewConversationLocal(ctx, chat1.ConversationIDTriple{
-		Tlfid:     tlfID,
-		TopicType: arg.TopicType,
-		// TopicID filled by server?
+	return conversations, nil
+}
+
+// getConversationInfo locates the conversation by using id, and returns with
+// all fields filled in conversationInfo, along with a ConversationIDTriple
+//
+// TODO: cache
+func (h *chatLocalHandler) getConversationInfo(ctx context.Context, id chat1.ConversationID) (conversationInfo keybase1.ConversationInfoLocal, triple chat1.ConversationIDTriple, err error) {
+	conversationRemote, err := h.remoteClient().GetConversationMetadataRemote(ctx, id)
+	if err != nil {
+		return conversationInfo, triple, err
+	}
+	if len(conversationRemote.MaxHeaders) == 0 {
+		return conversationInfo, triple, errors.New("empty conversation found")
+	}
+	// if no METADATA message exists, just use the newest one to get TLF Name
+	messageID := conversationRemote.MaxHeaders[0].MessageID
+	for _, header := range conversationRemote.MaxHeaders {
+		if header.MessageType == chat1.MessageType_METADATA {
```

Comment:
If a type of message doesn't exist, `MaxHeaders` wouldn't include it right? `messageID := conversationRemote.MaxHeaders[0].MessageID` makes sure we get a valid message ID, and the for loop after it tries to find one with METADATA type, and if found it, replace `messageID` with its ID